src/evo/mnhftx.cpp

```diff
@@ -77,9 +80,190 @@ bool CheckMNHFTx(const CTransaction& tx, const CBlockIndex* pindexPrev, TxValida
         return false;
     }
 
+    if (!Params().UpdateMNActivationParam(mnhfTx.signal.versionBit, pindexPrev->nHeight, pindexPrev->GetMedianTimePast(), true /* fJustCheck */)) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-mnhf-non-ehf");
+    }
+
+    return true;
+}
+
+static bool extractSignals(const CBlock& block, const CBlockIndex* const pindex, std::vector<uint8_t>& signals_to_process, BlockValidationState& state)
+{
+    AssertLockHeld(cs_main);
+
+    // we skip the coinbase
+    for (size_t i = 1; i < block.vtx.size(); ++i) {
+        const CTransaction& tx = *block.vtx[i];
+
+        if (tx.nVersion != 3 || tx.nType != TRANSACTION_MNHF_SIGNAL) {
+            // only interested in special TXs 'TRANSACTION_MNHF_SIGNAL'
+            continue;
+        }
+
+        TxValidationState tx_state;
+        if (!CheckMNHFTx(tx, pindex, tx_state)) {
+            return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, tx_state.GetRejectReason(), tx_state.GetDebugMessage());
+        }
+
+        MNHFTxPayload mnhfTx;
+        if (!GetTxPayload(tx, mnhfTx)) {
+            return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-tx-payload");
+        }
+        signals_to_process.push_back(mnhfTx.signal.versionBit);
+    }
+
+    // Checking that there's no any duplicates...
+    std::sort(signals_to_process.begin(), signals_to_process.end());
+    const auto it = std::unique(signals_to_process.begin(), signals_to_process.end());
+    if (std::distance(signals_to_process.begin(), it) != signals_to_process.size()) {
+        return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-duplicates");
+    }
+
+    return true;
+}
+
+bool CMNHFManager::ProcessBlock(const CBlock& block, const CBlockIndex* const pindex, bool fJustCheck, BlockValidationState& state)
+{
+    try {
+        std::vector<uint8_t> new_signals;
+        if (!extractSignals(block, pindex, new_signals, state)) {
+            // state is set inside extractSignals
+            return false;
+        }
+        if (new_signals.empty()) {
+            if (!fJustCheck) {
+                AddToCache(GetFromCache(pindex->pprev), pindex);
+            }
+            return true;
+        }
+
+        Signals signals = GetFromCache(pindex->pprev);
+        int mined_height = pindex->nHeight;
+
+        // Extra validation of signals to be sure that it can succeed
+        for (const auto& versionBit : new_signals) {
+            LogPrintf("%s: add mnhf bit=%d block:%s number of known signals:%lld\n", __func__, versionBit, pindex->GetBlockHash().ToString(), signals.size());
+            if (signals.find(versionBit) != signals.end()) {
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-duplicate");
```

Comment:
Say we activated some deployment via mnhf at some point. Will we be able to activate another deployment in the future via the same bit (after old deployment timeout)? It feels like signals never expire atm and we'll be getting `"bad-mnhf-duplicate"` forever for that bit.

Reply:
Firstly, you can't sign a signal with exactly same `version  bit` due to current quorum limitations. 
Same `version bit` -> `same requestID` -> message won't be signed again by any other quorum so far as I know.
so, you can't make a deployment with exactly same bit in future.

But it is not an issue for very long time. 
Firstly, the range of possible values EHF is 0...255 that is at least 200+ possible hard-forks in future (hunderds of major releases). It seems as we will have at least 50 years of future deployments for case of 4 hard-forks each years.

Secondly, so far as range of possible values EHF is bigger than range of version bits (0...28) and in case if we will continue old ways of hard-forking with miner's signaling, we can calculate bit for `Params().UpdateMNActivationParam()` through calculation of remainder to 29 (such as `signal bit % 29 = version bit`); but it requires DIP0023 to change, it seems as out-of-scope for current PR.

So far this approach seems fine.

Reply:
yes, indeed, [it works ](https://github.com/dashpay/dash/pull/5469/commits/2eaa67e4b0f4e776ae7736dd9f9deb330aaab7ac) if:
 - fix [restarting masternodes](https://github.com/dashpay/dash/pull/5469/commits/2eaa67e4b0f4e776ae7736dd9f9deb330aaab7ac#diff-1b5cd3f5c6ccc06db603a2f98d8b922d707b4f06fd51bc9e89973a3d556fd807R34)
 - [bump mocktime](https://github.com/dashpay/dash/pull/5469/commits/2eaa67e4b0f4e776ae7736dd9f9deb330aaab7ac#diff-1b5cd3f5c6ccc06db603a2f98d8b922d707b4f06fd51bc9e89973a3d556fd807R231) far in future.


WIP to implement expiring mined signals.

Reply:
@UdjinM6 signals can expire now: https://github.com/dashpay/dash/pull/5469/commits/6b7819f3b1ee332807e4b0fd6e9cad1327d50cae

Updates for functional tests to check it: https://github.com/dashpay/dash/pull/5469/commits/2eaa67e4b0f4e776ae7736dd9f9deb330aaab7ac